h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -3856,6 +3854,26 @@ public void setNonNegative() {
       }
     }
 
+    public void setNonNegative(Frame otherConst) {
+      List<String> constraintNames = extractVec2List(otherConst);
+      List<String> coeffNames = Arrays.stream(_dinfo.coefNames()).collect(Collectors.toList());
+      int numCoef = coeffNames.size();
+      for (int index=0; index<numCoef; index++) { // only changes beta constraints if not been specified before
+        if (!constraintNames.contains(coeffNames.get(index))) {  
+          _betaLB[index] = 0;
+          _betaUB[index] = Double.POSITIVE_INFINITY;
+        }
+      }
+    }
+    
+    public List<String> extractVec2List(Frame constraintF) {
+      List<String> constraintNames = new ArrayList<>();
+      int numRows = (int) constraintF.numRows();
+      for (int index=0; index<numRows; index++)
+        constraintNames.add(constraintF.vec(0).stringAt(index));
```

Comment:
In case we can afford it to handle it without MRTask, I would recommend to use [Vec.Reader.](https://github.com/h2oai/h2o-3/blob/master/h2o-core/src/main/java/water/fvec/Vec.java#L1197) The stringAt method is expensive, we need to find chunk for each time. Vec.Reader cache the chunk.

Reply:
Good idea.  Incorporated.